local-cluster/tests/local_cluster.rs

```diff
@@ -1465,6 +1474,220 @@ fn test_optimistic_confirmation_violation() {
     assert!(output.contains(&expected_log));
 }
 
+#[test]
+#[serial]
+#[ignore]
```

Comment:
Towards addressing this review comment: https://github.com/solana-labs/solana/pull/10718#discussion_r467340769

@carllin Well, I found it's very hard to make a validator violate optimistic confirmation if and only if tower is removed.
The hard part is that marking a slot as dead makes a validator to violate the optimistic conf. **even if tower is persited**. And eventually, the cluster makes new roots.

The initial situation is like this:
```
A: validator A (observer)
B: validator B (violator, will be restarted)

A & B:
0 -> 1 -> 2 -> ... -> 50 -> 51 (last vote for A and B; thus opt. confirmed)

```

Then mark 51 as dead only on B and B's new view after restart:

```
[...]  = not in bank_forks

B:
0 -> 1 -> 2 -> ... -> 50 [-> 51 (stray last vote for B)]

```

Even if we keep (optimistically-confirmed) slot 51 in towrer, due to forcibly marking 51 as dead, B's `bank_forks` doesn't have the ancestors info for slot 51. So, this falls into this clause: [`let last_vote_ancestors = ... { &empty_ancestors };`](https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR526). Then, [`!last_vote_ancestors.contains(lockout_interval_start)`](https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR604) always evaluates to `true`. So we can't exclude lockout for A correctly, which is actually  _ancestor of slot 51_. Finally, B then starts to vote creating a new fork like this:

```
B:
0 -> 1 -> 2 -> ... -> 49 -> 50 [-> 51 (stray last vote for B)]
                      |
                      +-> 56 -> 57
``` 

Ultimately, A starts to vote on the new bad fork, resulting in violated optimistic confirmation for the slot 51.

The forced dead slot 51 is the problem. But this shouldn't occur in practice. So this isn't bug per se.

But I couldn't come up with a good test scenario. Any suggestions?

Reply:
@ryoqun, ah I see! Does that mean we can remove this: https://github.com/solana-labs/solana/pull/10718/files#diff-a6e7be832768ca536ac5b7450a42ee09R1439-R1444 as well now?

Reply:
Hmm I couldn't come up with a simple example, but here's an interesting alternative idea

You have validator A + B with 31% and 36%  of the stake

0 -> 1 -> 2 -> 3 (A + B vote, optimistically confirmed)

Turn off A + B, start validator C with 33% of the stake with same ledger, but only up to slot 2. Have `C` generate some blocks like:

0 -> 1-> 2 -> 4

Then restart `A` which had 31% of the stake. With the tower, from `A`'s perspective it sees:

```
0 -> 1 -> 2 -> 3 (voted)
               |
                -> 4
```

So it should not be able to vote on `4`. Without the tower, `A` will vote on `4` b/c `C` has more stake and `B`'s vote for slot `3` is not observable (there's no descendant of `3` with the votes)

Reply:
oh this looks interesting case. Because this isn't critical part of this pr, I'll address this as a follow-up pr.

First let's land this 1700-lines pr to free me of rebasing... ;)

Reply:
yeah totally agree :smile: